### PR TITLE
Adding Cloud entitlement to ExportOptionPlist

### PR DESCRIPTION
**Pull request by meiravf**
When using the `auto` exportOption and the embedded provisioning profile contains the capability - `com.apple.developer.icloud-container-environment` the key `iCloudContainerEnvironment` needs to be added to the ExportOptionsPlist.

I have tested this change using `tfx` with different signing configurations (e.g. automatic, manual, developer cert, enterprise cert and app-store cert )

### CODE REVIEW

**Comment by meiravf** [comment]
@madhurig @hashtagchris Please review this PR.

File: Tasks/Xcode/Tests/L0.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import * as path from 'path';
 import * as assert from 'assert';
 import * as ttm from 'vsts-task-lib/mock-test';
+import { fail } from 'assert';
```
Comment (ranterle) [inline Tasks/Xcode/Tests/L0.ts]: Is this being used below?

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 2]
```diff
@@ -132,6 +132,52 @@ export async function findSigningIdentity(keychainPath: string) {
     }
 }
 
+/**
+ * Check if provisioning profile contains Cloud entitlement 
+ * @param provProfilePath
+ * @returns {boolean} 
+ */
+export async function cloudEntitlement(provProfilePath: string) {
```
Comment (lukekim) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: Should the name include a verb? E.g. `includesCloudEntitlement` or something like that? Also, if this returns a boolean use a return type? E.g. `) : Boolean {

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 3]
```diff
@@ -132,6 +132,52 @@ export async function findSigningIdentity(keychainPath: string) {
...
+    //find the provisioning profile details
+    let provProfileDetails: string;
+    let getProvProfileDetailsCmd: ToolRunner = tl.tool(tl.which('security', true));
```
Comment (lukekim) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: This could be `const`

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 4]
```diff
@@ -132,6 +132,52 @@ export async function findSigningIdentity(keychainPath: string) {
...
+    getProvProfileDetailsCmd.arg(['cms', '-D', '-i', provProfilePath]);
+    getProvProfileDetailsCmd.on('stdout', function (data) {
+        if (data) {
+            if (provProfileDetails) {
+                provProfileDetails = provProfileDetails.concat(data.toString().trim().replace(/[,\n\r\f\v]/gm, ''));
+            } else {
+                provProfileDetails = data.toString().trim().replace(/[,\n\r\f\v]/gm, '');
+            }
+        }
+    })
```
Comment (lukekim) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: Missing `;`

File: Tasks/Xcode/xcode.ts [hunk 5]
```diff
@@ -330,6 +330,7 @@ async function run() {
                 let exportOptionsPlist: string;
                 let archiveToCheck: string = archiveFolders[0];
                 let embeddedProvProfiles: string[] = tl.findMatch(archiveToCheck, '**/embedded.mobileprovision', { followSymbolicLinks: false, followSpecifiedSymbolicLink: false });
+                let cloudEntitlement: boolean;
```
Comment (ranterle) [inline Tasks/Xcode/xcode.ts]: Initializing this here would be helpful I think.

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 6]
```diff
@@ -132,6 +132,52 @@ export async function findSigningIdentity(keychainPath: string) {
...
+        fs.writeFileSync(tmpPlist, provProfileDetails);
+    } else {
+        throw tl.loc('ProvProfileDetailsNotFound', provProfilePath);
+    }
+
+    //use PlistBuddy to figure if cloud entitlement exists. 
+    let cloudEntitlement: string = await printFromPlist('Entitlements:com.apple.developer.icloud-container-environment', tmpPlist);
+
+    //delete the temporary plist file
+    let deletePlistCommand: ToolRunner = tl.tool(tl.which('rm', true));
```
Comment (ranterle) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: Can this be a `const`?

Comment (lukekim) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: Above you use `fs` to create the file, why use a ToolRunner to delete it? Can you not also use `fs` for that?

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 7]
```diff
@@ -132,6 +132,52 @@ export async function findSigningIdentity(keychainPath: string) {
...
+    if (provProfileDetails) {
+        //write the provisioning profile to a plist
+        tmpPlist = '_xcodetasktmp.plist';
+        fs.writeFileSync(tmpPlist, provProfileDetails);
+    } else {
+        throw tl.loc('ProvProfileDetailsNotFound', provProfilePath);
+    }
+
+    //use PlistBuddy to figure if cloud entitlement exists. 
+    let cloudEntitlement: string = await printFromPlist('Entitlements:com.apple.developer.icloud-container-environment', tmpPlist);
```
Comment (lukekim) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: `const`

File: Tasks/Xcode/Tests/L0.ts [hunk 8]
```diff
@@ -458,9 +459,94 @@ describe('Xcode L0 Suite', function () {
...
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.invokedToolCount == 21, 'Should have run \"PlistBuddy -c Add...\" four times, and 17 other command lines.');
+
+        done();
+    });
+
+    it('Xcode 9 signing defaults to manual, with auto export and cloud entitlement for production', (done: MochaDone) => {
+        this.timeout(1000);
+
+        let tp = path.join(__dirname, 'L0XCode9SigningDefaultsToAutoWithAutoExportAndCloudEntitlement.js');
```
Comment (ranterle) [inline Tasks/Xcode/Tests/L0.ts]: Maybe a `const`?

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 9]
```diff
@@ -132,6 +132,52 @@ export async function findSigningIdentity(keychainPath: string) {
...
+
+    //delete the temporary plist file
+    let deletePlistCommand: ToolRunner = tl.tool(tl.which('rm', true));
+    deletePlistCommand.arg(['-f', tmpPlist]);
+    await deletePlistCommand.exec();
+
+    if (cloudEntitlement) {
+        tl.debug('Provisioning Profile contains cloud entitlement');
```
Comment (lukekim) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: Is this debug command really required considering the return type? If so, why is the inverse (false) not also required?

File: Tasks/Xcode/Tests/L0.ts [hunk 10]
```diff
@@ -458,9 +459,94 @@ describe('Xcode L0 Suite', function () {
...
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.invokedToolCount == 22, 'Should have run \"PlistBuddy -c Add...\" four times, and 18 other command lines.');
+
+        done();
+    });
+
+    it('Xcode 9 signing defaults to manual, with auto export and cloud entitlement for development', (done: MochaDone) => {
+        this.timeout(1000);
+
+        let tp = path.join(__dirname, 'L0Xcode9ExportArchiveWithAutoAndCloudEntitlementForDevelopment.js');
```
Comment (ranterle) [inline Tasks/Xcode/Tests/L0.ts]: Can these be `const`s?

File: Tasks/Xcode/Tests/L0XCode9SigningDefaultsToAutoWithAutoExportAndCloudEntitlement.ts [hunk 11]
```diff
@@ -0,0 +1,188 @@
+
+import ma = require('vsts-task-lib/mock-answer');
+import tmrm = require('vsts-task-lib/mock-run');
+import path = require('path');
+import fs = require('fs');
+
+let taskPath = path.join(__dirname, '..', 'xcode.js');
```
Comment (ranterle) [inline Tasks/Xcode/Tests/L0XCode9SigningDefaultsToAutoWithAutoExportAndCloudEntitlement.ts]: Can these be `const`s?

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 12]
```diff
@@ -132,6 +132,52 @@ export async function findSigningIdentity(keychainPath: string) {
...
+            }
+        }
+    })
+    
+    await getProvProfileDetailsCmd.exec();
+
+    let tmpPlist: string;
+    if (provProfileDetails) {
+        //write the provisioning profile to a plist
+        tmpPlist = '_xcodetasktmp.plist';
```
Comment (lukekim) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: Where does this file get written? Should it not use a temporary or working directory? What happens if it already exists? Should it use some kind of random suffix?

File: Tasks/Xcode/Tests/L0Xcode9ExportArchiveWithAutoAndCloudEntitlementForDevelopment.ts [hunk 13]
```diff
@@ -0,0 +1,184 @@
+
+import ma = require('vsts-task-lib/mock-answer');
+import tmrm = require('vsts-task-lib/mock-run');
+import path = require('path');
+import fs = require('fs');
+
+let taskPath = path.join(__dirname, '..', 'xcode.js');
```
Comment (ranterle) [inline Tasks/Xcode/Tests/L0Xcode9ExportArchiveWithAutoAndCloudEntitlementForDevelopment.ts]: Can these be `const`s?

File: Tasks/Common/ios-signing-common/ios-signing-common.ts (hunk 2 above)
Comment (meiravf) [reply Tasks/Common/ios-signing-common/ios-signing-common.ts]: I wasn't sure what is the naming convention , I agree that a verb should be added. @madhurig @hashtagchris what are the naming convention in this case?

File: Tasks/Common/ios-signing-common/ios-signing-common.ts (hunk 12 above)
Comment (meiravf) [reply Tasks/Common/ios-signing-common/ios-signing-common.ts]: If it exists it will just overwrite it. 
The file is written in the local directory where the service is running.

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 14]
```diff
@@ -132,6 +132,53 @@ export async function findSigningIdentity(keychainPath: string) {
...
+        }
+    });
+    
+    await getProvProfileDetailsCmd.exec();
+
+    let tmpPlist: string;
+    if (provProfileDetails) {
+        //write the provisioning profile to a plist
+        tmpPlist = '_xcodetasktmp.plist';
+        tl.writeFile(tmpPlist,provProfileDetails);
```
Comment (madhurig) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: Missing space after the comma. You can try Shift+Option+F in VS Code on Mac to handle the formatting for you.

File: Tasks/Xcode/xcode.ts [hunk 15]
```diff
@@ -330,14 +330,16 @@ async function run() {
...
+                let cloudEntitlement: boolean = false;                    
+                
                 if (exportOptions === 'auto') {
                     // Automatically try to detect the export-method to use from the provisioning profile
                     // embedded in the .xcarchive file
                     if (embeddedProvProfiles && embeddedProvProfiles.length > 0) {
                         tl.debug('embedded prov profile = ' + embeddedProvProfiles[0]);
                         exportMethod = await sign.getProvisioningProfileType(embeddedProvProfiles[0]);
                         tl.debug('Using export method = ' + exportMethod);
+                        cloudEntitlement = await sign.includesCloudEntitlement(embeddedProvProfiles[0]);
```
Comment (madhurig) [inline Tasks/Xcode/xcode.ts]: It will be good if you can pass the exportMethod and the embeddedProvProfiles[0] and just get the cloud entitlement here. Change the method to getCloudEntitlement(), return Production/Development or empty string/null.

File: Tasks/Xcode/xcode.ts [hunk 16]
```diff
@@ -363,6 +365,14 @@ async function run() {
                     }
 
                     if (xcodeVersion >= 9 && exportOptions === 'auto') {
+                        if (cloudEntitlement === true) {
+                            tl.debug("Adding cloud entitlement");
+                            if (exportMethod === "app-store") {
```
Comment (madhurig) [inline Tasks/Xcode/xcode.ts]: This logic an move into the getCloudEntitlement helper and you can just check if(cloudEntitlement) and add the value to the exportOptionsPlist.

File: Tasks/Xcode/xcode.ts (hunk 15 above)
Comment (meiravf) [reply Tasks/Xcode/xcode.ts]: @madhurig updated, thanks!

File: Tasks/Common/ios-signing-common/ios-signing-common.ts [hunk 17]
```diff
@@ -132,6 +132,54 @@ export async function findSigningIdentity(keychainPath: string) {
...
 }
 
+    /**
+     * Get Cloud entitlement type Production or Development according to the export method - if entitlement doesn't exists in provisioning profile returns null 
+     * @param provisioningProfilePath
+     * @param exportMethod
+     * @returns {string} 
+     */
+export async function getCloudEntitlement(provisioningProfilePath: string, exportMethod: string): Promise<string> {
+    //find the provisioning profile details
```
Comment (hashtagchris) [inline Tasks/Common/ios-signing-common/ios-signing-common.ts]: I noticed we already repeat this code in several places. It should be factored out into a function. But that is a pre-existing issue I can address in another PR, especially if you're planning to port this fix back to 2.0.

File: Tasks/Xcode/Tests/L0.ts [hunk 18]
```diff
@@ -422,7 +422,7 @@ describe('Xcode L0 Suite', function () {
 
         assert(tr.stderr.length == 0, 'should not have written to stderr');
         assert(tr.succeeded, 'task should have succeeded');
-        assert(tr.invokedToolCount == 11, 'Should have run \"PlistBuddy -c Add...\" once, and 10 other command lines.');
+        assert(tr.invokedToolCount == 14, 'Should have run \"PlistBuddy -c Add...\" once, and 14 other command lines.');
```
Comment (hashtagchris) [inline Tasks/Xcode/Tests/L0.ts]: 13 :)

File: Tasks/Xcode/Tests/L0.ts [hunk 19]
```diff
@@ -458,9 +458,94 @@ describe('Xcode L0 Suite', function () {
             '-exportPath /user/build/_XcodeTaskExport_testScheme -exportOptionsPlist _XcodeTaskExportOptions.plist'),
             'xcodebuild exportArchive should have been run with -allowProvisioningUpdates to export the IPA from the .xcarchive');
 
+        assert(tr.ran("/usr/libexec/PlistBuddy -c Add iCloudContainerEnvironment string Production _XcodeTaskExportOptions.plist") === false,
+            'PlistBuddy add cloud entitlement should not run');
```
Comment (hashtagchris) [inline Tasks/Xcode/Tests/L0.ts]: People may overlook === false on the end; `!tr.ran` would be more obvious.

BTW, you typically don't need to check that something didn't run. I'd expect the test to fail if it exec'd a command line not listed in the exec section.

File: Tasks/Xcode/Tests/L0.ts [hunk 20]
```diff
@@ -458,9 +458,94 @@ describe('Xcode L0 Suite', function () {
...
+        assert(tr.succeeded, 'task should have succeeded');
+        assert(tr.invokedToolCount == 21, 'Should have run \"PlistBuddy -c Add...\" four times, and 17 other command lines.');
+
+        done();
+    });
+
+    it('Xcode 9 signing defaults to manual, with auto export and cloud entitlement for production', (done: MochaDone) => {
+        this.timeout(1000);
+
+        const tp = path.join(__dirname, 'L0XCode9SigningDefaultsToAutoWithAutoExportAndCloudEntitlement.js');
```
Comment (hashtagchris) [inline Tasks/Xcode/Tests/L0.ts]: ToAuto or ToManual? Test title and filename disagree.

File: Tasks/Xcode/xcode.ts [hunk 21]
```diff
@@ -363,6 +362,11 @@ async function run() {
                     }
 
                     if (xcodeVersion >= 9 && exportOptions === 'auto') {
+                        const cloudEntitlement = await sign.getCloudEntitlement(embeddedProvProfiles[0], exportMethod);
```
Comment (hashtagchris) [inline Tasks/Xcode/xcode.ts]: Is this relevant to Xcode 8 as well? If so, add a separate if block.

File: Tasks/Xcode/Tests/L0.ts (hunk 19 above)
Comment (meiravf) [reply Tasks/Xcode/Tests/L0.ts]: Right, removing this line.

File: Tasks/Xcode/xcode.ts (hunk 21 above)
Comment (meiravf) [reply Tasks/Xcode/xcode.ts]: Nope this is only relevant from Xcode 9 and after.

File: Tasks/Xcode/Tests/L0.ts (hunk 20 above)
Comment (meiravf) [reply Tasks/Xcode/Tests/L0.ts]: How about `L0Xcode9ExportArchiveWithAutoAndCloudEntitlementForProduction`?

**Comment by meiravf** [comment]
@hashtagchris fixed , thanks!

**Comment by smeirsha** [comment]
@hashtagchris  I don't have permission to merge the PR, can you please do it.

**Comment by meiravf** [comment]
@hashtagchris I don't have permission to merge the PR can you do it please.
